v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/PipelineController.java

```diff
@@ -0,0 +1,238 @@
+import com.google.cloud.teleport.v2.source.reader.io.jdbc.iowrapper.JdbcIoWrapper;
+import com.google.cloud.teleport.v2.source.reader.io.row.SourceRow;
+import com.google.cloud.teleport.v2.source.reader.io.schema.SourceSchema;
+import com.google.cloud.teleport.v2.source.reader.io.transform.ReaderTransform;
+import com.google.cloud.teleport.v2.spanner.ddl.Ddl;
+import com.google.cloud.teleport.v2.spanner.migrations.exceptions.InvalidOptionsException;
+import com.google.cloud.teleport.v2.spanner.migrations.schema.ISchemaMapper;
+import com.google.cloud.teleport.v2.spanner.migrations.schema.IdentityMapper;
+import com.google.cloud.teleport.v2.spanner.migrations.schema.SessionBasedMapper;
+import com.google.cloud.teleport.v2.spanner.migrations.shard.Shard;
+import com.google.cloud.teleport.v2.spanner.migrations.spanner.SpannerSchema;
+import com.google.cloud.teleport.v2.transformer.SourceRowToMutationDoFn;
+import com.google.cloud.teleport.v2.writer.DeadLetterQueue;
+import com.google.cloud.teleport.v2.writer.SpannerWriter;
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.stream.Collectors;
+import org.apache.beam.repackaged.core.org.apache.commons.lang3.StringUtils;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.PipelineResult;
+import org.apache.beam.sdk.coders.SerializableCoder;
+import org.apache.beam.sdk.io.gcp.spanner.MutationGroup;
+import org.apache.beam.sdk.io.gcp.spanner.SpannerConfig;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTagList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Place perform pipeline level orchestration, scheduling and tuning operations. */
+public class PipelineController {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SourceDbToSpanner.class);
+
+  static PipelineResult executeSingleInstanceMigration(
+      SourceDbToSpannerOptions options, Pipeline pipeline, SpannerConfig spannerConfig) {
+
+    Ddl ddl = SpannerSchema.getInformationSchemaAsDdl(spannerConfig);
+    ISchemaMapper schemaMapper = PipelineController.getSchemaMapper(options, ddl);
+
+    List<String> tablesToMigrate =
+        PipelineController.listTablesToMigrate(options.getTables(), schemaMapper, ddl);
+    // Read data from source
+    ReaderImpl reader =
+        ReaderImpl.of(
+            JdbcIoWrapper.of(
+                OptionsToConfigBuilder.MySql.configWithMySqlDefaultsFromOptions(
+                    options, tablesToMigrate, null)));
+    migrateForReader(options, pipeline, spannerConfig, ddl, schemaMapper, reader, "");
+    return pipeline.run();
+  }
+
+  private static void migrateForReader(
```

Comment:
PLease add a comment mentioning this instantiates an end-to-end pipeline give a db source.

Reply:
Done.